### PR TITLE
T18940 test case id view

### PR DESCRIPTION
**Pull request by gctucker**
Add a view to show all the details for a given test case result.

This includes all the relevant test case meta-data (kernel revision, compiler, device type...) as well as measurements, log lines associates with the test case and regression history.

### CODE REVIEW

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+        /* ToDo: highlight current revision row, and show "future" regressions
+         * when looking at other revisions (needs database schema changes).
+         */
+        results.regressions.reverse().forEach(function(data) {
+            var row = document.createElement('tr');
+            var createdOn = new Date(data.created_on.$date);
+            var cell;
+
+            cell = row.insertCell(-1);
+            cell.appendChild(document.createTextNode(data.kernel));
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Maybe could be interesting to add a link in kernel column to `test/job/branch/kernel`.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+                createdOn.getSeconds()
+            ));
+
+            cell = row.insertCell(-1);
+            cell.setAttribute('class', 'pull-center');
+            cell.appendChild(ttable.statusNode(data.status));
+
+            cell = row.insertCell(-1);
+            cell.setAttribute('class', 'pull-center');
+            cell.appendChild(ttable.detailsNode(
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Details has a link just to "refresh" the page going to /test/case/id/<id>.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+        var jobLink;
+        var describeNode;
+        var buildsLink;
+        var createdOn;
+        var dateNode;
+        var status;
+
+        job = results.job;
+        kernel = results.kernel;
+        translatedURI = urls.createFileServerURL(gFileServer, results);
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: in the case of `updateCaseDetails` translatedURI is not used and I think could be removed. Maybe it could be used to redirect to pages where you can see details from device, arch or soc..

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+            results.boot_log, results.boot_log_html, results.lab_name,
+            translatedURI[0], translatedURI[1]);
+
+        defconfigNode = html.tooltip();
+        defconfigLink = document.createElement('a');
+        defconfigLink.appendChild(
+            document.createTextNode(results.defconfig_full));
+        defconfigPath = translatedURI[1] + "/kernel.config";
+        defconfigLink.href =
+            translatedURI[0].path(defconfigPath).normalizePath().href()
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: Links created by `updateGroupDetails` for: Defconfig URL, Kernel image and device tree don't work. Will new views be created or will you make adjustments redirecting to pages that we already have?

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+            cell = row.insertCell(-1);
+            cell.appendChild(document.createTextNode(
+                createdOn.toDateString('%Y')
+            ));
+
+            cell = row.insertCell(-1);
+            cell.appendChild(document.createTextNode(
+                createdOn.getHours() + ':' +
+                createdOn.getMinutes() + ':' +
+                createdOn.getSeconds()
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: is there any special reason to not use logic from `utils/table` to build table?

**Review by alexandrasp — CHANGES_REQUESTED** [review]

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 1 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Yes, although I didn't want to confuse things.  I would expect people to be able to the previous runs of the same test case, not to land on a page with all the tests run for kernel revision.  So this is something that could be improved but I would rather keep it simple for now and only have the "details" link which is unambiguous.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 2 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: What I wanted to do was to have the same table visible on the pages showing previous runs of the test, with the full regression history including "future" runs, and highlight the row matching the current view.  But this is proving a bit difficult with the current schema.  I guess we could still have something instead of the "details" link to show it is the current view, I'll think about it.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 3 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Yes, I'm removing `translatedURI` from `updateCaseDetails()`.  This was left when I reworked the code and addedd `updateGroupDetails()`.  Maybe we can add more links, but I think that would be part of a follow-up set of improvements, taking into account the whole user experience across the website.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 4 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: It works as far as I can tell, see this page for example:
https://staging.kernelci.org/test/case/id/5e94a5b59c954596db219e58/

Can you please provide a link to a page where it doesn't work?

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 5 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: `utils/table` are used to create [DataTables](https://datatables.net/) whereas I only needed a plain table in this case.  But I think it's worth trying with DataTables, at least to enable paging and to deal with cases where there are many regression rows such as this one:
https://staging.kernelci.org/test/case/id/5e94a5b59c954596db219e5d/

So I'll take a look.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 4 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: take a look on this page: https://staging.kernelci.org/test/case/id/5e963b0edb5c1a46d9b8e8f9/ 
 
I saw that it works for most cases, but..

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 3 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Done in latest version.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 5 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: OK I've rewritten this with DataTables now in the latest version.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 4 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: Great, thanks for spotting this.  It looks like other test results with the same build are fine, so that's a strange corner-case.  I'll take a look tomorrow.

**Comment by gctucker** [comment]
Tested OK on staging with the new table with regression history:
https://staging.kernelci.org/test/case/id/5e965bd80d9262b2939f201a/

Next one to fix is the defconfig/kernel links on this view:
https://staging.kernelci.org/test/case/id/5e963b0edb5c1a46d9b8e8f9/
the URL is:
https://storage.staging.kernelci.org/kernelci/staging-20200414.0/arm64-defconfig/kernel.config
instead of:
https://storage.staging.kernelci.org/kernelci/staging.kernelci.org/staging-20200414.0/arm64/defconfig/gcc-8/kernel.config
Probably something to do with the `file_server_resource` somehow.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 4 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: Found the root cause, it's an issue with sub-groups missing some values.  Here's a PR to fix it in the backend:
https://github.com/kernelci/kernelci-backend/pull/232

We'll see in tomorrow's staging data if this is resolved.  It worked with my local set-up.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 2 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Now the "details" links has been removed with the move to DataTables, each row has a link to a test case view.  I think the "current" row could have a different class to show with a different background or something like that.  I'll take a look if something simple can be done, but I don't think it's a blocker for this PR.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 4 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:198]: There was one more issue with the`modules_dir`, this is actually not used anywhere so I've updated the code to look for `modules` only instead.  That worked:
https://staging.kernelci.org/test/case/id/5e97e876e1e1cb6b7888cbb0/

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+    var gCaseId;
+    var gFileServer;
+    var gRegressionTable;
+
+    setTimeout(function() {
+        document.getElementById('li-test').setAttribute('class', 'active');
+    }, 15);
+
+    function detailsFailed() {
+        html.replaceByClassTxt('loading-content', '?');
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js:44]: I think we can remove `detailsFailed` and do that html replace into some generic function that could be used by get from group, case and regression.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                render: _renderStatus,
+            },
+        ]
+
+        gRegressionTable
+            .data(data)
+            .columns(columns)
+            .languageLengthMenu('Revisions per page')
+            .lengthMenu([10, 25, 50, 75, 100])
+            .rowURL('/test/case/id/%(test_case_id)s')
```
Comment (alexandrasp) [inline app/dashboard/static/js/app/view-tests-case-id.2020.3.js:322]: it still a little bit weird to me a link that just "reload" the page, but it is not a big problem. I see.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 2 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: Maybe we can have another view to show something specific related to that test case, but..I don't know what could be useful for kernel developers, maybe more detailed log in case of fail(s). I am just thinking :}

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 6 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:44]: Maybe, something part of the next round of improvements.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 7 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:322]: It would be more confusing to not have the current revision in the table.

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 2 above)
Comment (gctucker) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js]: This view is the one to show all the details for a single test case.

**Review by alexandrasp — APPROVED** [review]

File: app/dashboard/static/js/app/view-tests-case-id.2020.3.js (hunk 7 above)
Comment (alexandrasp) [reply app/dashboard/static/js/app/view-tests-case-id.2020.3.js:322]: Maybe we can have a view to show something that could be interesting for kernel developers related to a test case in the next round of improvements. I am just thinking, I don't know what kind of thing could be useful.. :}
